### PR TITLE
Add Metric Emitter, which right now emits a metric once a minute for true replication lag in nanoseconds.

### PR DESCRIPTION
**Pull request by ZackLK**

### CODE REVIEW

File: service/history/replication/metrics_emitter.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+func (m *metricsEmitterImpl) determineReplicationLatency(remoteClusterName string) (time.Duration, error) {
+	logger := m.logger.WithTags(tag.RemoteCluster(remoteClusterName))
+	lastReadTaskID := m.shardData.GetClusterReplicationLevel(remoteClusterName)
+	maxReadLevel := m.shardData.GetTransferMaxReadLevel()
+
+	// TODO I should probably use task_store instead to get from cache, even though I don't need it hydrated
+	// TODO Do I need to use a different context? Or add a deadline / cancellation?
+	tasks, _, err := m.reader.Read(ctx.Background(), lastReadTaskID, maxReadLevel)
+	if err != nil {
```
Comment (vytautas-karpavicius) [inline service/history/replication/metrics_emitter.go:165]: Correct, you do not need this hydrated. I think we can leave it as is for now.
`context.Background()` should be fine. I do not see much benefit on restricting deadline for such a background processing task.

File: service/history/replication/metrics_emitter.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+)
+
+const (
+	metricsEmissionInterval = time.Minute
+)
+
+type (
+	// MetricsEmitter is responsible for emitting replication metrics occasionally.
+	// TODO: Should this be merged into task_processor or is this a good split?
+	MetricsEmitter interface {
```
[the remark this answers is not in the document]
Comment (ZackLK) [reply service/history/replication/metrics_emitter.go:47]: Only purpose was to allow easier unit testing.

File: service/history/replication/metrics_emitter.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+func NewMetricsEmitter(
+	shardData MetricsEmitterShardData,
+	reader taskReader,
+	metricsClient metrics.Client,
+) *metricsEmitterImpl {
+	shardID := shardData.GetShardID()
+	currentCluster := shardData.GetClusterMetadata().GetCurrentClusterName()
+
+	remoteClusters := shardData.GetClusterMetadata().GetRemoteClusterInfo()
+	remoteClusterNames := make([]string, len(remoteClusters))
```
[the remark this answers is not in the document]
Comment (ZackLK) [reply service/history/replication/metrics_emitter.go:84]: Oops, good point!

File: service/history/replication/metrics_emitter.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+		scope.UpdateGauge(metrics.ReplicationLatency, float64(replicationLatency.Nanoseconds()))
+		logger.Debug(fmt.Sprintf("ReplicationLatency metric emitted: %v", float64(replicationLatency.Nanoseconds())))
+	}
+}
+
+func (m *metricsEmitterImpl) determineReplicationLatency(remoteClusterName string) (time.Duration, error) {
+	logger := m.logger.WithTags(tag.RemoteCluster(remoteClusterName))
+	lastReadTaskID := m.shardData.GetClusterReplicationLevel(remoteClusterName)
+	maxReadLevel := m.shardData.GetTransferMaxReadLevel()
```
[the remark this answers is not in the document]
Comment (ZackLK) [reply service/history/replication/metrics_emitter.go:160]: Yeah, good point. I initially thought of doing +1, but then thought that would be an inefficient read from the database...

File: service/history/replication/metrics_emitter.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		return
+	}
+
+	m.logger.Info("ReplicationMetricsEmitter shutting down.")
+	close(m.done)
+}
+
+func (m *MetricsEmitterImpl) emitMetricsLoop() {
+	ticker := time.NewTicker(metricsEmissionInterval)
+	defer ticker.Stop()
```
Comment (davidporter-id-au) [inline service/history/replication/metrics_emitter.go:122]: Does this need a panic handler?

**Review by davidporter-id-au — APPROVED** [review]

**Review by vytautas-karpavicius — APPROVED** [review]

File: service/history/replication/metrics_emitter.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	}
+}
+
+func (m *MetricsEmitterImpl) determineReplicationLatency(remoteClusterName string) (time.Duration, error) {
+	logger := m.logger.WithTags(tag.RemoteCluster(remoteClusterName))
+	lastReadTaskID := m.shardData.GetClusterReplicationLevel(remoteClusterName)
+
+	tasks, _, err := m.reader.Read(ctx.Background(), lastReadTaskID, lastReadTaskID+1)
+	if err != nil {
+		logger.Error("Error reading", tag.Error(err))
```
Comment (vytautas-karpavicius) [inline service/history/replication/metrics_emitter.go:156]: Perhaps a more descriptive error message would be helpful here?

**Pull request merged**